### ISSUE TITLE
a bug in groupBy v2?

### ISSUE THREAD

**Issue by himanshug** [druid-io/druid#3983]
One of my users (using version 0.9.2) is using time extraction function in a groupBy query.

results are always correct with groupBy-v1 while with groupBy-v2 they are sometimes correct and sometimes not 
however, without the extractionFn in the query groupBy-v2 seems to always produce correct result.

here is what the query looks like...

```
{
  "queryType": "groupBy",
  "dataSource": {
    "type": "table",
    "name": "xxx"
  },
  "intervals": {
    "type": "intervals",
    "intervals": ["2016-03-01T00:00:00.000Z/2016-03-02T00:00:00.000Z", "2016-04-01T00:00:00.000Z/2016-04-02T00:00:00.000Z", "2016-05-01T00:00:00.000Z/2016-05-02T00:00:00.000Z"]
  },
  "filter": {
    "type": "and",
    "fields": [{
      "type": "or",
      "fields": [{
        "type": "selector",
        "dimension": "myDate",
        "value": "20160301"
      }, {
        "type": "selector",
        "dimension": "myDate",
        "value": "20160401"
      }, {
        "type": "selector",
        "dimension": "myDate",
        "value": "20160501"
      }]
    }, {
      "type": "selector",
      "dimension": "xx",
      "value": "xx"
    }, {
      "type": "or",
      "fields": [{
        "type": "selector",
        "dimension": "xx",
        "value": "yy"
      }, {
        "type": "selector",
        "dimension": "xx",
        "value": "yy"
      }, {
        "type": "selector",
        "dimension": "xx",
        "value": "yy"
      }]
    }]
  },
  "granularity": {
    "type": "all"
  },
  "dimensions": [{
    "type": "extraction",
    "dimension": "myDate",
    "outputName": "Month",
    "extractionFn": {
      "type": "time",
      "timeFormat": "yyyyMMdd",
      "resultFormat": "yyyyMM"
    }
  }, {
    "type": "default",
    "dimension": "xx",
    "outputName": "xx"
  }],
  "aggregations": [{
    "type": "longSum",
    "name": "xx",
    "fieldName": "xx"
  }],
  "postAggregations": [],
  "limitSpec": {
    "type": "default",
    "columns": [{
      "dimension": "Month",
      "direction": "ascending",
      "dimensionOrder": {
        "type": "lexicographic"
      }
    }],
    "limit": 1000
  },
  "context": {
    "groupByStrategy": "v2",
    "groupByIsSingleThreaded": true
  },
  "descending": false
}
```

**Comment by himanshug** [comment]
@gianm @jon-wei have you guys seen something similar in 0.9.2 and if there was a bug that got fixed in 0.10.0 ?

**Comment by pavanab4u** [comment]
### result with v2 where month is totally wrong
```
[
    {
        "version": "v1",
        "timestamp": "2016-03-01T00:00:00.000Z",
        "event": {
            "Month": "000001",
            "xx": 14,
            "yy": "D"
        }
    },
    {
        "version": "v1",
        "timestamp": "2016-03-01T00:00:00.000Z",
        "event": {
            "Month": "201604",
            "xx": 1,
            "yy": "S"
        }
    },
    {
        "version": "v1",
        "timestamp": "2016-03-01T00:00:00.000Z",
        "event": {
            "Month": "471301",
            "xx": 1,
            "yy": "S"
        }
    },
    {
        "version": "v1",
        "timestamp": "2016-03-01T00:00:00.000Z",
        "event": {
            "Month": "471301",
            "xx": 1,
            "yy": "T"
        }
    }
]
```

**Comment by patelh** [comment]
+1

**Comment by gianm** [comment]
@himanshug when you say "sometimes correct and sometimes not" what is the correct result and what is the incorrect result you get? How many segments are you querying over? Is caching enabled anywhere?

**Comment by himanshug** [comment]
@pavanab4u can you answer @gianm questions ?

**Comment by gianm** [comment]
SimpleDateFormat is used by TimeDimExtractionFn but is not thread-safe according to http://www.icu-project.org/apiref/icu4j/com/ibm/icu/text/SimpleDateFormat.html. groupBy v2 definitely applies this extractionFn from different threads since extractionFns are generally assumed to be thread safe. Maybe this is the problem.

I'm not sure why groupBy v1 doesn't exhibit this problem, maybe for some reason it's less likely to apply extraction functions concurrently.

Could you try hacking up a build that synchronizes `apply` on TimeDimExtractionFn? If that fixes it then that confirms this is the problem.

**Comment by patelh** [comment]
```
"dimensions": [{
    "type": "extraction",
    "dimension": "myDate",
    "outputName": "Month",
    "extractionFn": {
      "type": "time",
      "timeFormat": "yyyyMMdd",
      "resultFormat": "yyyyMM"
    }
  }
```

For the above, the correct result would be the result of the field which timeFormat yyyyMMdd converted to yyyyMM.  e.g. 20170228 to 201702 , and with v1 the result is already as expected, with v2, randomly we see corrupted result, e.g. values such as 000001 or 471301 and etc...

**Comment by himanshug** [comment]
@gianm that theory might be correct . I notice that they're using `"groupByIsSingleThreaded": true` which means v1 query would not do merging in parallel but v2 ignores that flag.
let me see if it can be reproduced locally first... deploying custom release on user's production cluster is kinda hard.

**Comment by himanshug** [comment]
@gianm ok once I remove `"groupByIsSingleThreaded": false` , then even v1 groupBy starts producing funny looking results, so most likely TimeExtractionFunction is buggy under concurrency.

**Comment by himanshug** [comment]
https://github.com/druid-io/druid/pull/3984 should probably fix this, will need to further deploy it on user cluster to be somewhat sure.

**Comment by patelh** [comment]
OMG, so groupByIsSingleThreaded is no longer supported by v2 ?  This will impact # of concurrent user queries.

**Comment by patelh** [comment]
What about priority ?

**Comment by gianm** [comment]
Priority is supported by all query types including groupBy v2.

Single threaded merging isn't supported by groupBy v2. Could you expand on how that will affect you? I didn't implement single threaded merging in v2 because I guessed multithreaded merging would generally be better than single threaded with the new engine.

**Comment by patelh** [comment]
Yes it would be better.  But if I need to have higher QPS, where the # of segments scanned are arbitrary for every query, the # of requests that can be processed would go down.  By guaranteeing only a single thread is used, I can control the kind of queries I want to use more than a single thread.  This allows for better control of resources across use cases / customers where the SLA is may be different across those use cases.

**Comment by gianm** [comment]
Thanks, that makes sense. Query priorities should be able to accomplish something similar (they'll let a higher priority query take over the processing thread pool). The only downside of priorities should be if each individual segment scan takes a long time, which would mean it might take too long for a higher priority query to take over. In that case it does make sense to limit the number of threads used by lower priority groupBy queries. (or make the segment scans take less time…)

If query priorities aren't enough and thread # control is essential, you could set your groupBy engine back to v1 through config, or you could do a patch giving groupBy v2 the ability to limit thread use too.

With either engine please keep in mind that memory use can still be an issue -- if you have a lot of thread-count-limited groupBy queries then even though they won't be taking up too many threads, they may be taking up a lot of memory. groupBy v2 controls this through numMergeBuffers which ends up being the max number of queries you can run concurrently.

**Comment by patelh** [comment]
Yea priority would be one way to do something similar.  But in the case where I have multiple queries at the same priority, there is no way to guarantee resources to each of those queries since they'd be processed in order (IIRC), where as if we could control the number of threads for each, we may be able to process them concurrently.

@himanshug can we please add this to our list to add groupBySingleThreaded to v2.

We'll continue to use v1 until we have groupBySingleThreaded in v2 for those use cases where we currently use groupBySingleThreaded.

**Comment by himanshug** [comment]
@gianm @patelh given that users depend on singleThreaded flag to manage the resources , we'll restore that flag for groupBy v2 .

**Comment by gianm** [comment]
> But in the case where I have multiple queries at the same priority, there is no way to guarantee resources to each of those queries since they'd be processed in order (IIRC), where as if we could control the number of threads for each, we may be able to process them concurrently.

By default the query processing runs in no particular order, but you can set `druid.processing.fifo` to do queries in fifo order (see also https://metamarkets.com/2016/impact-on-query-speed-from-forced-processing-ordering-in-druid/).

If you want to run all active queries of the same priority concurrently rather than fifo, you could probably do that by adding an option like `druid.processing.fifo` that does some different behavior.

This is just as an FYI in case you want to explore this path. But if you think your needs are satisfied by just making groupBys single threaded, then that works too.
